js/spreadsheet.js

```diff
@@ -68,244 +68,196 @@ var spreadsheet = function(options) {
+  /**
+   * The function that helps us to find end index for further selection
+   * 
+   * @param {array} tableData - an array of data from the Handsontable library
+   * @param {int} startSearchFrom - index of the row/col from what we should start a search
+   * @param {int} colIndex - a not necessary option that tells us what we need to look, a row index or column index
+   * @returns {int} - index of the row or column
+   * 
+   */
+  function getEndIndex(tableData, startSearchFrom, colIndex) {
+    var currentIndex = startSearchFrom;
 
-      // Looking for the last row in the array of allData
-      // When we got a null value in the cell it means that we reached the range borders
-      for (var i = firstRow; i < allData.length; i++) {
-        if (allData[i][firstCol] === null) {
-          lastRow = i;
+    if (currentIndex || currentIndex === 0) {
+      for (currentIndex; currentIndex < tableData.length; currentIndex++) {
+        if (tableData[currentIndex][colIndex] === null) {
           break;
         }
       }
-
-      lastRow = _.max([lastRow - 1, 0]);
-    } else if (moveTo.right) {
-      // When data located on the right of the selected cell
-      firstCol = startAt[1];
-
-      for (var i = firstCol + 1; i < allData.length; i++) {
-        if (allData[startAt[0]][i] === null) {
-          lastCol = i - 1;
+    } else {
+      for (currentIndex; currentIndex < tableData.length; currentIndex++) {
+        if (tableData[currentIndex] === null) {
           break;
         }
       }
+    }
 
-      for (var i = startAt[0]; i > 0; i--) {
-        if (allData[i][lastCol] === null) {
-          firstRow = i ? i - 1 : i;
-        }
-      }
+    return currentIndex;
+  }
 
-      firstRow = firstRow || 0;
+  /**
+   * Function to get coordinates which we need to select
+   * 
+   * @param {Array} cellPosition - an array of the selected coordinates
+   * 
+   * @returns {Array/Boolean} - coordinates that need to be selected. Example of the returned data: [[startRow, startCol, endRow, endCol]]
+   */
+  function coordinatesToSelect(cellPosition) {
+    var firstCol, lastCol, firstRow, lastRow, tableData;
+
+    // Returns array of the data from the table with Handsontable API
+    tableData = hot.getData();
+    selectedCellPosition = {
+      row: cellPosition[0][0],
+      col: cellPosition[0][1]
+    };
 
-      for (var i = firstRow; i < allData.length; i++) {
-        if (allData[i][lastCol] === null) {
-          lastRow = i - 1;
-          break;
-        }
-      }
-    } else if (moveTo.top) {
-      // When data located on the top of the selected cell
-      lastRow = startAt[0];
+    var nearestDataAt = nearestCellsWithData(selectedCellPosition, tableData);
 
-      for (var i = lastRow - 1; i > 0; i--) {
-        if (allData[i][startAt[1]] === null) {
-          firstRow = i;
-          break;
-        }
-      }
+    // Return false in the case when we need to select all table data
+    if (nearestDataAt.selectAll) {
```

Comment:
Move the `for (var key in ...)` logic from inside the `getAdjacentCellsWithData()` function to here.